Add pgtype.Uint64 to support pgtype.PgLSN type (closes #1407)

hnakamur opened the pull request:
Discussed in [Add pgtype.Uint64 to support pg_lsn type · Issue #1407 · jackc/pgx](https://github.com/jackc/pgx/issues/1407)

jackc commented:
The tests of `pg_lsn` need to skip CockroachDB for CI.

I also don't see how the text format could be working. `TestConnQuerPgLSN` is not testing encoding at all and is only testing decoding binary. That test should be in the `pgtype` package and use `pgxtest.RunValueRoundTripTests`.

jackc commented:
Looks like Cockroach is still getting triggered somewhere.

I ran the tests locally and it seems to work -- though TBH I still don't see exactly how. The text format is different for pg_lsn than a normal integer and I don't see a scan plan or an encode plan to the text format. I guess `ScanText` and `TextValue` or doing the job. I suppose it works, but that's not the normal path.

Also, I noticed that the database/sql implementations for `Value()` are returning `uint64`. This is not in spec (only allowed types are listed here: https://pkg.go.dev/database/sql/driver#Value).

hnakamur commented:
> Looks like Cockroach is still getting triggered somewhere.

It was `TestConnQuerPgLSN` in query_test.go. I removed it at https://github.com/jackc/pgx/pull/1416/commits/349d80d3c72471584b3c23c0102f5e51a325c001.

> I ran the tests locally and it seems to work -- though TBH I still don't see exactly how. The text format is different for pg_lsn than a normal integer and I don't see a scan plan or an encode plan to the text format. I guess ScanText and TextValue or doing the job. I suppose it works, but that's not the normal path.

Yes, I added debug logs and confirmed `ScanText` and `TextValue` are called.
It's my first time to try to add a type to pgx, and honestly I'm not familiar with the normal path.
Could you point me to an example or document that I can learn the normal path?

> Also, I noticed that the database/sql implementations for Value() are returning uint64. This is not in spec (only allowed types are listed here: https://pkg.go.dev/database/sql/driver#Value).

Thanks I didn't know that. I changed `Value()` to return int64 instead of uint64 at https://github.com/jackc/pgx/pull/1416/commits/99e8ae54befd89e68919ab7ab1ed3fc447f5d5d0.
I also changed `Scan()` to handle `int64` as well.

jackc commented:
> Could you point me to an example or document that I can learn the normal path?

I would expect the LSN codec to define its own `PlanScan` and `PlanEncode`. They would reuse the binary format code from uint64 but use their own paths for the text format. Check out `jsonb` code. It has a similar relationship to `json` that `pg_lsn` has the `uint64`. The only difference is that for json/jsonb it is the binary format that differs and for pg_lsn/uint64 it is the text format that differs.

> Thanks I didn't know that. I changed Value() to return int64 instead of uint64 at https://github.com/jackc/pgx/commit/99e8ae54befd89e68919ab7ab1ed3fc447f5d5d0.
I also changed Scan() to handle int64 as well.

Sorry to keep nitpicking, but that won't work in all cases either. What happens if the number is greater than the max int64? It will need to encode the value into a string to be safe.

hnakamur commented:
> I would expect the LSN codec to define its own PlanScan and PlanEncode. They would reuse the binary format code from uint64 but use their own paths for the text format. Check out jsonb code. It has a similar relationship to json that pg_lsn has the uint64. The only difference is that for json/jsonb it is the binary format that differs and for pg_lsn/uint64 it is the text format that differs.

Thanks for your explanation. I rewrote PgLSNCodec at https://github.com/jackc/pgx/pull/1416/commits/5ffe8df4adf9e82cb02fe53c874f767e53090f48.

> Sorry to keep nitpicking, but that won't work in all cases either. What happens if the number is greater than the max int64? It will need to encode the value into a string to be safe.

Thanks. I deleted `Scan()` since `TestPgLSNCodec` passes without it.

jackc commented on pgtype/pg_lsn.go:
```diff
@@ -0,0 +1,206 @@
+			}
+		}
+	}
+
```
I don't think this block is necessary here. That is a special case for json.

jackc commented on pgtype/uint64_test.go:
```diff
@@ -0,0 +1,21 @@
+)
+
+func TestUint64Codec(t *testing.T) {
+	pgxtest.RunValueRoundTripTests(context.Background(), t, defaultConnTestRunner, pgxtest.KnownOIDQueryExecModes, "pg_lsn", []pgxtest.ValueRoundTripTest{
```
This needs to skip CockroachDB as well.

But it also makes me wonder if this is actually testing `Uint64Codec`. It's using the `pg_lsn` type. That would mean that `PgLSNCodec` is being used. We'd need to use another real uint64 PG type for this test... Hmm... That makes me wonder again about whether Uint64 should actually exist. I looked in the  PostgreSQL `pg_type` table and your example of `FullTransactionId` isn't exposed there. I don't think there are any other uint64 types in PostgreSQL. (Run `select * from pg_type where typlen = 8;` in PostgreSQL to see all candidates).

Unless we can find another uint64 type in PostgreSQL I think pg_lsn is all that should exist.

hnakamur replied:
> I don't think there are any other uint64 types in PostgreSQL. (Run select * from pg_type where typlen = 8; in PostgreSQL to see all candidates).

Thanks. I looked at source code of `macaddr8` and `xid8` and confirmed that they are not backed with `uint64`.

> Unless we can find another uint64 type in PostgreSQL I think pg_lsn is all that should exist.

I am convinced now and am closing this pull request and the issue [Add pgtype.Uint64 to support pg_lsn type · Issue #1407 · jackc/pgx](https://github.com/jackc/pgx/issues/1407).

Sorry for taking your time and thanks again for your review.
